src/FtpExtractor.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\FtpExtractor;
+
+use Keboola\Component\UserException;
+use League\Flysystem\FileNotFoundException;
+use League\Flysystem\Filesystem as FtpFilesystem;
+use Symfony\Component\Filesystem\Filesystem;
+use Webmozart\Glob\Glob;
+
+class FtpExtractor
+{
+    private const RECURSIVE_COPY = true;
+    private const FTP_FILETYPE_FILE = 'file';
+    private const FILE_DESTINATION_KEY = 'destination-path';
+    private const FILE_TIMESTAMP_KEY = 'timestamp';
+    private const FILE_SOURCE_KEY = 'source-path';
+
+    /**
+     * @var FtpFilesystem
+     */
+    private $ftpFilesystem;
+
+    /**
+     * @var bool
+     */
+    private $onlyNewFiles;
+
+    /**
+     * @var bool
+     */
+    private $isWildcard;
+
+    /**
+     * @var array
+     */
+    private $filesToDownload;
+
+    public function __construct(bool $onlyNewFiles, bool $isWildcard, FtpFilesystem $ftpFs)
+    {
+        $this->ftpFilesystem = $ftpFs;
+        $this->onlyNewFiles = $onlyNewFiles;
+        $this->isWildcard = $isWildcard;
+        $this->filesToDownload = [];
+    }
+
+    public function copyFiles(string $sourcePath, string $destionationPath, FileStateRegistry $registry): int
+    {
+        if ($this->isWildcard) {
+            $this->prepareToDownloadFolder($sourcePath, $destionationPath);
+        } else {
+            $this->prepareToDownloadSingleFile($sourcePath, $destionationPath);
+        }
+        return $this->download($registry);
+    }
+
+    private function prepareToDownloadFolder(string $sourcePath, string $destinationPath): void
+    {
+        $basePath = Glob::getBasePath(GlobValidator::convertToAbsolute($sourcePath));
+        try {
+            $items = $this->ftpFilesystem->listContents($basePath, self::RECURSIVE_COPY);
+        } catch (\LogicException $e) {
```

Comment:
urcite by to nemelo konvertovat LogicException na UserException, spis bych rekl, ze by to melo zachytavat League\Flysystem\Exception;

jestli je potreba konvertovat FilesystemNotFoundException na user exception, tak radsi at jsou tam dva catche nez catch na logicException

Reply:
Souhlas, taky neuznávám zachytávání catch(\Exception $e), ale tady ta knihovna vyhazuje nějaký vyjímky, který nejsou v jejím namespace :-1: https://github.com/thephpleague/flysystem/blob/master/src/Adapter/Ftp.php#L137
Buďto jim můžu udělat pull request a přetypovat vyjímky nebo zachytávat tyhle "root" vyjímky.

Reply:
Není to krásné, ale je to řešení: https://github.com/keboola/ex-teradata/blob/master/src/ExceptionHandler.php

Reply:
@tomasfejfar  jo to bude funkční :+1: 
Co kdybych tam zatím nechal tyhle hnusný catch bloky a udělal issue na nějaký hezčí zachytávání? Myslím, že 99% budou špatný credentials a do tý doby bych udělal nějaký jiný sexy řešení?

Mimochodem FTP adapter při bad credentials vyhazuje LogicException a SFTP RuntimeException :-1:

Reply:
uh, tak to je mazec :( Jsem pro nechat tam klidne hnusny try-catch bloky, ale at jsou na to testy - bad credentials a bad server pro ftp a sftp a nenalezeny soubor - to jsou veci, ktere jdou dobre otestovat a bude s tim nejvic problemu